### PR TITLE
Manual close end point now allows to close manually once

### PR DESCRIPTION
**Pull request by marta-lokhova**
This isn't well tested yet (and probably not simplified either), but opening a preliminary PR so that I can do some simulation testing.

### CODE REVIEW

**Comment by graydon** [comment]
Looks fairly good but I am a little concerned with a few points:
  - The names "once" and "all" are a little uninformative / non-obvious to users, and passing a boolean flag named `once` that we check the _negation_ of seems a little hard to follow; especially when combined with the fact that "all" needs the flag to be set, and `once` needs FORCE_SCP to _not_ be set. Overall it just feels confusing. This is subjective and you're welcome to tell me "this is the best that can be done!" but it's sorta a gut sense I get reading it, that I'll have a hard time editing the code in the future / telling what it's supposed to do in which combinations.
  - More worryingly: "once" can be called more than once, and I'm not sure it's actually safe to call Herder::bootstrap more than once. Maybe? Might also cause havoc.
  - I'm not 100% clear on why you need to restoreScpState in the non-FORCE_SCP path of bootstrap. What's different about it? 
  - Overall I suspect the close-once flag might serve a subtly-different purpose than the `MANUAL_CLOSE` flag. If I understand correctly, `MANUAL_CLOSE` was added as a kind of escape hatch to let running networks that have "lost leadership" get forced back into operation by an operator. This is, I believe, why it is also not hidden behind `BUILD_TESTS`: so that production nodes can be nudged back into SCP when they have given up on it or are all waiting for one another. This might no longer be necessary at all (given the changes you landed in https://github.com/stellar/stellar-core/pull/2612) but if it is true (or if we want to keep it around for purposes of backward compatibility with scripts, or mere paranoia) then I think that's different than what you're adding here. I think what you're adding here is a one-time, idempotent trigger that _negates_ `--wait-for-consensus` mode and says ok, stop waiting now. If that's true I might suggest:
    - Move `--wait-for-consensus` behind `BUILD_TESTS` if we think nobody needs to do this in production (I think .. not?)
    - Rename `Config.FORCE_SCP` to `Config.WAIT_FOR_CONSENSUS` to reduce the number of double negatives. Limit the residue of the `FORCE_SCP` name to the deprecation warnings and back-compat config option parser. 
    - Add a new HTTP method called, say, `bootstrap` that _is_ behind `BUILD_TESTS` that calls bootstrap iff `WAIT_FOR_CONSENSUS` is true and bootstrap hasn't happened yet. If bootstrap has happened, return an error indicating that.

How does that sound to you?

**Comment by marta-lokhova** [comment]
Thanks for taking a look! I made changes to address the first two bullets, I agree that keeping track of all the flags was confusing.

Regarding "restoreScpState" in bootstrap: when we startup with --wait-for-consensus, we're not in sync technically, but `ledgerClosed` expects herder to be in sync. So with this change we basically pretend like we got in sync, and closed ledger, so we can trigger a new one.

For the last bullet, I haven't implemented all of your suggestions yet, as I realized that I'm confused by the existing manual_close functionality. So I'd like to hash out some questions on that first. 
My main issue is that I don't quite understand how MANUAL_CLOSE is supposed to be used, as the documentation seems to contradict the implementation.
* `manualclose` endpoint says that it forces ledger close. In reality, I don't think it's true -- nodes trigger a consensus round. When MANUAL_CLOSE is set to true, nodes discard scp messages from peers, so I don't understand how it is possible to close a ledger for networks that consist of more than 1 node.
* I don't think manualclose was intended to be a production endpoint. `commands.md` and `Application::manualClose` both claim that it is for testing only.

So I guess my question is: do we want to keep manualclose as is, and just have `bootstrap` endpoint like you suggested, or do we want to re-visit manualclose, and change its behavior to the new bootstrapping logic? (@graydon @MonsieurNicolas )

**Comment by graydon** [comment]
Fine questions! I await @MonsieurNicolas input as I do not have a clear sense of the answers myself, clearly :)

File: src/main/CommandHandler.cpp [hunk 1]
```diff
@@ -104,6 +104,7 @@ CommandHandler::CommandHandler(Application& app) : mApp(app)
     addRoute("generateload", &CommandHandler::generateLoad);
     addRoute("testacc", &CommandHandler::testAcc);
     addRoute("testtx", &CommandHandler::testTx);
+    addRoute("bootstrap", &CommandHandler::bootstrap);
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp:107]: should we always expose this? This seems to be useful in general to spin up nodes, so maybe shouldn't be limited to test builds?

File: src/herder/HerderImpl.cpp [hunk 2]
```diff
@@ -125,14 +125,25 @@ HerderImpl::getStateHuman() const
...
-    assert(getSCP().isValidator());
-    assert(mApp.getConfig().FORCE_SCP);
+    if (!mBootstrapped)
+    {
+        mBootstrapped = true;
+
+        CLOG(INFO, "Herder") << "Force joining SCP with local state";
+        assert(getSCP().isValidator());
+
+        if (!mApp.getConfig().FORCE_SCP)
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:135]: Should we just always restore SCP state in the code path `mHerder->restoreState()` and not do it arbitrarly late: the only difference between `FORCE_SCP` and no `FORCE_SCP` will then only be that we trigger a ledger  or not.

File: src/herder/HerderImpl.cpp [hunk 3]
```diff
@@ -125,14 +125,25 @@ HerderImpl::getStateHuman() const
...
+        {
+            auto const& lcl = mLedgerManager.getLastClosedLedgerHeader();
+            mHerderSCPDriver.restoreSCPState(lcl.header.ledgerSeq,
+                                             lcl.header.scpValue);
+        }
 
-    mLedgerManager.bootstrap();
-    mHerderSCPDriver.bootstrap();
+        mLedgerManager.bootstrap();
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:142]: I am really not sure this is right.
we're introducing a race between whatever that node is doing and calling `bootstrap`: those bootstrap calls may reset a bunch of things that the node was actually doing.

I thought the less dangerous version of this PR was to just allow `manualclose` to be called at any time?

File: src/herder/HerderImpl.cpp (hunk 2 above)
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:135]: I did that initially, but then moved "restore state" path to later for the following reasons:
* If we move this to `mHerder->restoreState()`, we'd need to setup a timer to properly go out of sync as `restoreState` is used for normal node operation as well. For the simulation purposes, I don't think we want that, as nodes should stay "in sync" until manualclose is called (which can take arbitrarily long) in order to trigger the next ledger 
* minor: there is an existing optimization to not go in sync when we're at the genesis ledger and waiting for consensus  

I could change it as per your suggestion, but it seems like this way we get more flexibility. What do you think?

**Comment by marta-lokhova** [comment]
That makes sense @MonsieurNicolas. I think I initially started with triggering the next ledger instead of bootstrapping, but as you pointed out, there is some prep required to bring the node into proper state to trigger, and bootstrap was already doing some of it. I'll update the code to check timers, then get in sync, then call `triggerNextLedger`.

File: src/herder/HerderImpl.cpp [hunk 4]
```diff
@@ -825,6 +825,36 @@ HerderImpl::getMaxSeqInPendingTxs(AccountID const& acc)
...
+    // middle of a consensus round. Either way, we do not want
+    // to trigger ledger, as the node is already making progress
+    if (mTriggerTimer.seq() > 0)
+    {
+        CLOG(DEBUG, "Herder") << "Skipping setInSyncAndTriggerNextLedger: "
+                                 "trigger timer already set";
+        return;
+    }
+
+    // Bring Herder and LM in sync in case they aren't
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:841]: Yikes, it ends up being not as straightforward as I thought 😅 

As far as I can tell, `restoreSCPState` doesn't do what you want when lcl is genesis (so starting fresh) as it no-ops in that case.

More importantly, this code is still unsafe, imagine the following sequence takes place:
1. node waits for consensus
2. node receives enough messages to externalize
3. (node may transition to catchup)
4. node goes out of sync
5. we invoke the `manualclose` endpoint

At this point
* `restoreSCPState` will perform potentially unsafe operations (overrides a bunch of SCP related fields, potentially rolls back certain things in SCP itself)
* `mLedgerManager.bootstrap` potentially sets the wrong state (if catching up)

Now, do we need to actually do this?
* for `mLedgerManager`
   * if it's "catching up" we have to no-op
   * otherwise, we can call `mLedgerManager.bootstrap()` (that will either no-op or perform the BOOTING->SYNCED transition) - maybe that function should be renamed `forceTransitionToSynced`?
* for SCP, the only thing we care about is to make sure that we're nominating against the local state (so `lcl`) so that it can properly validate values. I don't think we need to mess with "tracking"
   * basically in this particular situation, we can bypass the check `!mHerderSCPDriver.trackingSCP()` ; we still need to ensure that `mLedgerManager.isSynced()` is true though
   * we could call a variation of `restoreSCP` on startup, but that is a minor improvement. This would make it safer if we're using this endpoint when restarting the network from a state that's not genesis
       * we'd remove the condition on `FORCE_SCP` (ie, always restore when not on genesis)
       * make it no-op if the node is not validating
       * always restores the SCP state
       * skips setting tracking state when `FORCE_SCP` is not set (doesn't call `HerderSCPDriver::restoreSCPState`)

Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:841]: Updated.

As discussed offline, I did not change `restoreSCPState`, as it already handles non-genesis ledgers properly.

**Comment by MonsieurNicolas** [comment]
r+ b513585af87a110624ccb79f461a0457a1839931

**Comment by MonsieurNicolas** [comment]
@latobarita: retry
